### PR TITLE
Rewrite the wrapper to typescript.

### PR DESCRIPTION
**Pull request by jansiegel**
- Rewritten the rollup config
- Rewritten the wrapper to typescript

To test the final build, run:
```
npm run build && npm run copy-scopes
```

This will create the `build` directory with both `ce` and `pro` builds and a merged typescript definition file (`index.d.ts`).

Related issue: #14

### CODE REVIEW

File: src/common/index.tsx [hunk 1]
```diff
@@ -1,6 +1,14 @@
 import React from 'react';
 import Handsontable from 'hot-alias';
-import {SettingsMapper} from './settingsMapper';
+import { SettingsMapper } from './settingsMapper';
+
+export interface Props extends Handsontable.DefaultSettings {
+  data?: any;
```
Comment (wojciechczerniak) [inline src/common/index.tsx]: Can we be more precise here? `any[][] | object[]` ?

File: .babelrc [hunk 2]
```diff
@@ -13,4 +13,4 @@
       ]
     }
   }
-}
```
Comment (swistach) [inline .babelrc]: Could you add an empty line below?

File: .config/base.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import replace from 'rollup-plugin-replace';
+import typescript from 'rollup-plugin-typescript2';
+
+const envHotType = process.env.HOT_TYPE;
+
+export const baseConfig = {
+  input: 'src/common/index.tsx',
+  plugins: [
+    replace({
+      'hot-alias': envHotType === 'pro' ? 'handsontable-pro' : 'handsontable',
```
Comment (swistach) [inline .config/base.js]: Maybe `process.env.HOT_TYPE` could be just `handsontable-pro` or `handsontable`?

File: .config/commonjs.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+const filename = 'react-handsontable.js';
+
+export const cjsConfig = {
+  output: {
+    format: env,
+    indent: false,
+    file: './commonjs/' + envHotType + '/' + filename
+  },
+  plugins: baseConfig.plugins.concat([commonjs()])
+};
```
Comment (swistach) [inline .config/commonjs.js]: Could you add an empty line below?

File: .config/es.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+          declaration: true
+        }
+      }
+    }),
+    babel({
+      exclude: 'node_modules/**',
+    }),
+    nodeResolve(),
+  ],
+};
```
Comment (swistach) [inline .config/es.js]: Could you add an empty line below?

File: .config/factory.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+       Object.assign(config[key], newConfig[key]);
+
+     } else {
+       config[key] = newConfig[key];
+     }
+    }
+  }
+
+  return config;
+}
```
Comment (swistach) [inline .config/factory.js]: Could you add an empty line below?

File: .config/factory.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  const newConfigs = {
+    cjs: cjsConfig,
+    es: esConfig,
+    umd: umdConfig,
+    min: minConfig,
+  };
+  const newConfig = newConfigs[env];
+
+  for ( let key in newConfig ) {
+    if (newConfig.hasOwnProperty(key)) {
```
Comment (swistach) [inline .config/factory.js]: Maybe it should be more like in our helper? https://github.com/handsontable/handsontable/blob/master/src/helpers/object.js#L228

File: .config/helpers/licenseBanner.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  const envHotType = process.env.HOT_TYPE;
+  const packageBody = require('./src/' + envHotType + '/package.json');
+
+  let licenseBody = fs.readFileSync(path.resolve(__dirname, './LICENSE'), 'utf8');
+  licenseBody += '\nVersion: ' + packageBody.version + ' (built at ' + new Date().toString() + ')';
+
+  config.output.banner = '/*!\n' + licenseBody.replace(/^/gm, ' * ') + '\n */';
+
+  return config;
+}
```
Comment (swistach) [inline .config/helpers/licenseBanner.js]: Could you add an empty line below?

File: .config/helpers/licenseBanner.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+export function addLicenseBanner(config) {
+  const path = require('path');
+  const fs = require('fs');
+  const envHotType = process.env.HOT_TYPE;
+  const packageBody = require('./src/' + envHotType + '/package.json');
```
Comment (swistach) [inline .config/helpers/licenseBanner.js]: Could you use
```
const packageBody = require(['./src/', envHotType, '/package.json'].join(''));
// or
const packageBody = require(`./src/${envHotType}/package.json`);
```
?
Below in the same way.

File: .config/minified.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        unsafe_comps: true,
+        warnings: false
+      }
+    })
+  ])
+};
+
+addLicenseBanner(minConfig);
+
+export {minConfig};
```
Comment (swistach) [inline .config/minified.js]: Could you add an empty line below?

File: .config/umd.js [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    format: env,
+    name: 'Handsontable.react',
+    indent: false,
+    file: './dist/' + envHotType + '/' + filename
+  }
+};
+
+addLicenseBanner(umdConfig);
+
+export { umdConfig };
```
Comment (swistach) [inline .config/umd.js]: Could you add an empty line below?

File: rollup.config.js [hunk 12]
```diff
@@ -1,88 +1,3 @@
...
-
-if (envHotType === 'ce') {
-  config.external.push('handsontable');
-
-} else if (envHotType === 'pro') {
-  config.external.push('handsontable-pro');
-}
-
-export default config;
+export default createConfig();
```
Comment (swistach) [inline rollup.config.js]: Could you add an empty line below?

File: src/common/index.tsx [hunk 13]
```diff
@@ -1,6 +1,14 @@
 import React from 'react';
 import Handsontable from 'hot-alias';
-import {SettingsMapper} from './settingsMapper';
+import { SettingsMapper } from './settingsMapper';
+
+export interface Props extends Handsontable.DefaultSettings {
```
Comment (swistach) [inline src/common/index.tsx]: I think it should be separated file. Would you mind to add a description? What it is at all.

File: src/common/index.tsx [hunk 14]
```diff
@@ -24,45 +32,56 @@ import {SettingsMapper} from './settingsMapper';
...
+   */
+  style: React.CSSProperties;
 
   /**
    * Set the reference to the main Handsontable DOM element.
    *
    * @param {HTMLElement} element The main Handsontable DOM element.
    */
-  setHotElementRef(element) {
+  private setHotElementRef(element: HTMLElement) {
```
Comment (swistach) [inline src/common/index.tsx]: Could you add `private setHotElementRef(element: HTMLElement): void` ?

File: src/common/index.tsx (hunk 13 above)
Comment (swistach) [reply src/common/index.tsx]: By the way, maybe this interface should be named as `HotTableProps`?

File: src/common/settingsMapper.ts [hunk 15]
```diff
@@ -1,8 +1,16 @@
 import Handsontable from 'hot-alias';
+import { Props } from './index';
 
 export class SettingsMapper {
+
+  // TODO: this should be removed after the release of Handsontable 5.0.0+. It's a workaround for a faulty Handsontable definition file.
+  private hooks: any;
+
```
Comment (swistach) [inline src/common/settingsMapper.ts]: This empty line isn't necessary.

File: .config/base.js (hunk 3 above)
Comment (jansiegel) [reply .config/base.js]: Why?

File: .config/commonjs.js (hunk 4 above)
Comment (jansiegel) [reply .config/commonjs.js]: It has an empty last line, github just doesn't show it.

File: .config/es.js (hunk 5 above)
Comment (jansiegel) [reply .config/es.js]: It has an empty last line, github just doesn't show it.

File: .config/factory.js (hunk 6 above)
Comment (jansiegel) [reply .config/factory.js]: It has an empty last line, github just doesn't show it.

File: .config/helpers/licenseBanner.js (hunk 8 above)
Comment (jansiegel) [reply .config/helpers/licenseBanner.js]: It has an empty last line, github just doesn't show it.

File: .config/umd.js (hunk 11 above)
Comment (jansiegel) [reply .config/umd.js]: It has an empty last line, github just doesn't show it.

File: rollup.config.js (hunk 12 above)
Comment (jansiegel) [reply rollup.config.js]: It has an empty last line, github just doesn't show it.

File: .config/factory.js (hunk 7 above)
Comment (jansiegel) [reply .config/factory.js]: Is there a reason for how it's made in the helper?

File: .babelrc (hunk 2 above)
Comment (swistach) [reply .babelrc]: :+1:

File: src/common/settingsMapper.ts (hunk 15 above)
Comment (jansiegel) [reply src/common/settingsMapper.ts]: I'll keep it to separate the workaround from the "actual code", it'll be easier to notice after releasing HOT 5.0.0+.

File: src/common/index.tsx [hunk 16]
```diff
@@ -1,6 +1,14 @@
...
 import Handsontable from 'hot-alias';
-import {SettingsMapper} from './settingsMapper';
+import { SettingsMapper } from './settingsMapper';
+
+export interface Props extends Handsontable.DefaultSettings {
+  data?: any[][] | object[];
+  id?: string,
+  className?: string,
+  style?: React.CSSProperties,
+  settings?: Handsontable.DefaultSettings
```
Comment (wojciechczerniak) [inline src/common/index.tsx]: Just for the sake of discussion. I've noticed that in angular wrapper we're using `GridSettings` which is basically the same as `DefaultSettings` but looks more like a type than an instance. `DefaultSettings` are "THE settings". Probably doesn't matter that much, but would be cool to be consistent across wrappers too. What do you think @jansiegel  @swistach ?

Comment (jansiegel) [reply src/common/index.tsx]: If I'm reading this correctly, `DefaultSettings` extends `GridSettings` with... nothing, so they really are the same thing. But I think using `DefaultSettings` makes more sense, as it potentially _might_ be extended with something someday (and we will probably want to be able to make use of the new properties).

File: src/common/settingsMapper.ts (hunk 15 above)
Comment (swistach) [reply src/common/settingsMapper.ts]: :+1:

**Pull request merged**